### ISSUE TITLE
ISCC pre-processor command line options bug

### ISSUE THREAD

**Issue by extensia** [jrsoftware/issrc#170]
There appears to be a bug in the command line handling for ISCC.exe. When executed from a directory other than the directory containing the ISCC executable, pre-processor options such as /p are not recognized or shown in the command line help.

Looking at the command line handling in the source code, I see that the program is checking for the presence of ISPP.dll in the same directory as ISCC.exe to decided whether to enable the pre-processor features. This check appears to be broken for some non-obvious reason. Below is the relevant line from the application source code.

[ISCC.dpr line 596](https://github.com/jrsoftware/issrc/blob/master/Projects/ISCC.dpr#L596)

    IsppMode := FileExists(ExtractFilePath(NewParamStr(0)) + 'ispp.dll');

**Comment by DeeHants** [comment]
I'm not 100% sure in this case, but the first parameter (`NewParamStr(0)`) is generally the executing application's path, meaning the current folder doesn't come into play.

It also works fine on my machine:

```
H:\>"c:\Program Files (x86)\Inno Setup 5\ISCC.exe" /?
Inno Setup 5 Command-Line Compiler
Copyright (C) 1997-2014 Jordan Russell. All rights reserved.
Portions Copyright (C) 2000-2014 Martijn Laan
Inno Setup Preprocessor
Copyright (C) 2001-2004 Alex Yackimoff. All rights reserved.
```

**Comment by DeeHants** [comment]
What is the full path to your `ISCC.exe` and is there the associated `ispp.dll`?

**Comment by martijnlaan** [comment]
Works fine for me as well. Only after deleting ispp.dll the ISPP options weren't shown in the command line help.

**Comment by extensia** [comment]
I looked into the issue in more detail and it is a bit more complicated than I initially thought. First a bit of background. I have developed a custom Gradle plugin that downloads a zip-packaged copy of Inno Setup from our artifact repository (Artifactory) and uses it to compile an install script as part of a larger build process. This works well since we are effectively doing an XCOPY deploy of Inno Setup into the build directory which avoids the need to install Inno Setup on our TeamCity build agents.

After receiving your response, I fired up Sysinternals Process Explorer and can see that Inno Setup is not looking for ISPP.dll in the directory where ISCC.exe is located as expected. Instead, Inno Setup is looking in the standard Inno Setup install directory located my development machine, but only when Inno Setup is invoked without a path to the executable. Running "ISCC.exe" from the XCOPY deploy folder shows all options as expected, but is in fact finding ISPP.dll in the standard install directory. When run with a path (e.g. C:\MyProject\Build\InnoSetup\ISCC.exe), the ISPP.dll is not found and the options are not displayed.

Looking at the ISCC.dpr file it is not clear why this is happening. I may attempt to debug the problem in Delphi, but the oldest version I have installed at the moment is XE7. Will Inno Setup compile clealy with XE7?

Thanks,
David

**Comment by extensia** [comment]
After a bit more testing I believe the bug is related to handling of drive letters. Part of the wierdness I was seeing appears to have been related to the standard Inno Setup directory being in my system path.

Here are some simple test cases that demonstrate the nature of the problem I am seeing:

Test Setup
1) Manually copy the Inno Setup folder to a work directory on C: (e.g. C:\InnoTest)
2) Manually copy the Inno Setup folder to a work directory on E: (e.g. E:\InnoTest)
3) Rename the folder created by the Inno Setup installer so it is not available (e.g. add an "_")

Test Case #1
1) C:
2) cd C:\InnoTest
3) ISCC.exe  (OK - displays pre-processor options)

Test Case #2
1) C:
2) cd \
3) C:\InnoTest\ISCC.exe  (OK - displays pre-processor options)

Test Case #3
1) E:
2) cd E:\InnoTest
3) ISCC.exe  (FAIL - missing pre-processor options)

Test Case #4
1) E:
2) cd \
3) E:\InnoTest\ISCC.exe  (FAIL - missing pre-processor options)

Test Case #5
1) E:
2) cd \
3) C:\InnoTest\ISCC.exe  (OK - displays pre-processor options)

There is clearly some sensitivity to the drive letter where the executing ISCC.exe is located that is causing the pre-processor DLL to not be found. What is puzzling is that Process Monitor clearly shows that the ISPP.dll file is being enumerated and found by the OS:

"5:20:51.4377414 PM","ISCC.exe","13592","QueryDirectory","E:\InnoSetup","SUCCESS","0: .., 1: Compil32.exe, 2: Default.isl, 3: inno-setup-version.txt, 4: isbunzip.dll, 5: isbzip.dll, 6: ISCC.exe, 7: Iscmplr.dll, 8: islzma.dll, 9: islzma32.exe, 10: islzma64.exe, 11: <b>ISPP.dll</b>, 12: ISPPBuiltins.iss, 13: isscint.dll, 14: isunzlib.dll, 15: iszlib.dll, 16: Languages, 17: license.txt, 18: Setup.e32, 19: Setupldr.e32, 20: WizModernImage-IS.bmp, 21: WizModernImage.bmp, 22: WizModernSmallImage-IS.bmp, 23: WizModernSmallImage.bmp"

**Comment by extensia** [comment]
I believe the mystery is solved. The problem I am seeing is because <b>the copy of ISCC.exe that I have been using is apparently lacking the pre-processor logic!</b>

I managed to recompile ISCC.exe with a copy of Delphi 5 from a old laptop backup image and observed that the new build worked correctly. Comparing the executable installed by ispack-5.5.5.exe with my copy I noticed that the version I have been using is substantially smaller (117,760 bytes v.s. 159,232 bytes). Presumably this means that the version of ISCC.exe I have been using for testing is a different build than the one installed by ispack-5.5.5.exe.

The root cause of this issue appears to be that <b>the v5.5.5 copy of ISCC.exe that was checked into the GitHub jrsoftware/ispack project is not the same as was used for the ispack-5.5.5.exe installer</b>. Looking at the installer more closely I noticed that it reports itself as 5.5.5(a). Perhaps my problem is related to the issue that caused the 5.5.5(a) build to be released?

I see that 5.5.6 has now been released and will update to that version. A lot of time wasted tracking down this issue, but that just goes to prove the old saying about what happens when one makes assumptions!

Best Regards,
David

**Comment by martijnlaan** [comment]
There wasn't just one ISCC before. You said:

> v5.5.5 copy of ISCC.exe that was checked into the GitHub jrsoftware/ispack project is not the same
> as was used for the ispack-5.5.5.exe installer

but actually this ISCC.exe is the non ISPP ISCC. Besides that there's also ISPPCC.exe which is the ISPP ISCC. See https://github.com/jrsoftware/ispack/tree/is-5_5_5/isfiles to see both files. ISPPCC is also mentioned in https://github.com/jrsoftware/issrc/blob/is-5_5_5/README.md

But none of this applies anymore to 5.5.6 since ISPPCC has been merged into ISCC. See #115.

Secondly, there's no second 5.5.5 build. The version it reports itself as is '5.5.5 (a)' not '5.5.5(a)'. As documented in the help file the (a) means you've installed the ANSI version, see http://jrsoftware.org/ishelp/topic_unicode.htm.

Well, hopefully this clear things up and still looking forward to pull requests from you :)

**Comment by extensia** [comment]
I must be blind...over a bit over worked...since I missed the ISPPCC executable in the repo. I suspected that might be the case, but just didn't see it when I looked. Perhaps because the name is all lower case ;)

The reason for using the ispack repo is because I needed to repackage the Inno Setup binaries as an Apache Maven compatible artifact. I have a gradle script that automatically extracts the required files from the ispack sources, packages them and uploads them to our internal Artifactory server. Our production builds pull down the specified version of Inno Setup on demand for use in automated builds.
